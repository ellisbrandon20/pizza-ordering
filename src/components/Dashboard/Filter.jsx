import { Anchor, Button, Container, Div, Dropdown, Input } from "atomize";
import { Component } from "react";

const CrustDropdownList = ({ updateFilter, updateShowDropDown, showCrustDropdown }) => (
	<Div>
		{['Original', 'Thin', 'Stuffed', 'Deepdish', 'Clear'].map((name, i) => (
			<Anchor
				key={`crust-dropdown-${i}`}
				d='block'
				p={{ y: '0.25rem' }}
				onClick={() => {
					updateFilter('filCrust', name === 'Clear' ? undefined : name);
					updateShowDropDown('showCrustDropdown', !showCrustDropdown);
				}}
			>
				{name}
			</Anchor>
		))}
	</Div>
);

const FlavorDropdownList = ({ updateFilter, updateShowDropDown, showFlavorDropdown }) => (
	<Div>
		{['Cheese', 'Pepperoni', 'Sausage', 'Supreme', 'Clear'].map((name, i) => (
			<Anchor
				key={`flavor-dropdown-${i}`}
				d='block'
				p={{ y: '0.25rem' }}
				onClick={() => {
					updateFilter('filFlavor', name === 'Clear' ? undefined : name);
					updateShowDropDown('showFlavorDropdown', !showFlavorDropdown);
				}}
			>
				{name}
			</Anchor>
		))}
	</Div>
);

const SizeDropdownList = ({ updateFilter, updateShowDropDown, showSizeDropdown }) => (
	<Div>
		{[
			{ label: 'Small (12")', value: 'S' },
			{ label: 'Medium (16")', value: 'M' },
			{ label: 'Large (18")', value: 'L' },
			{ label: 'Clear', value: undefined },
		].map((name, i) => (
			<Anchor
				key={`size-dropdown-${i}`}
				d='block'
				p={{ y: '0.25rem' }}
				onClick={() => {
					updateFilter('filSize', name.value);
					updateShowDropDown('showSizeDropdown', !showSizeDropdown);
				}}
			>
				{name.label}
			</Anchor>
		))}
	</Div>
);

export class Filter extends Component {
	constructor(props) {
		super(props);

		this.state = {
			'showCrustDropdown': false,
			'showFlavorDropdown': false,
			'showSizeDropdown': false,
		};
	}

	updateShowDropDown = (filDropdown, value) => {
		this.setState({ [filDropdown]: value });
	};

	clearAllFilters = () => {
		this.props.updateFilter('filCrust', undefined);
		this.props.updateFilter('filFlavor', undefined);
		this.props.updateFilter('filSize', undefined);
		this.props.updateFilter('filTableNo', NaN);
		document.querySelector('input').value = undefined;
	}

	render() {
		const { showCrustDropdown, showFlavorDropdown, showSizeDropdown } = this.state;
		const { filCrust, filFlavor, filSize } = this.props;

		return (
			<Container
				w="50rem"
				m={{ l: 'auto', r: 'auto', t: '1.25rem' }}
			>
				<Div d='flex'>
					<Dropdown
						isOpen={showCrustDropdown}
						onClick={() => this.updateShowDropDown('showCrustDropdown', !showCrustDropdown)}
						menu={<CrustDropdownList updateFilter={this.props.updateFilter} updateShowDropDown={this.updateShowDropDown} showCrustDropdown={showCrustDropdown} />}
						w='8rem'
						m={{ l: '0.5rem', r: '0.5rem' }}
						borderColor="danger500"
						focusBorderColor="danger700"
						hoverBorderColor="danger700"
					>
						{filCrust === undefined ? 'Crust' : filCrust}
					</Dropdown>
					<Dropdown
						isOpen={showFlavorDropdown}
						onClick={() => this.updateShowDropDown('showFlavorDropdown', !showFlavorDropdown)}
						menu={<FlavorDropdownList updateFilter={this.props.updateFilter} updateShowDropDown={this.updateShowDropDown} showFlavorDropdown={showFlavorDropdown} />}
						w='8rem'
						m={{ l: '0.5rem', r: '0.5rem' }}
						borderColor="danger500"
						focusBorderColor="danger700"
						hoverBorderColor="danger700"
					>
						{filFlavor === undefined ? 'Flavor' : filFlavor}

					</Dropdown>
					<Dropdown
						isOpen={showSizeDropdown}
						onClick={() => this.updateShowDropDown('showSizeDropdown', !showSizeDropdown)}
						menu={<SizeDropdownList updateFilter={this.props.updateFilter} updateShowDropDown={this.updateShowDropDown} showSizeDropdown={showSizeDropdown} />}
						w='8rem'
						m={{ l: '0.5rem', r: '0.5rem' }}
						borderColor="danger500"
						focusBorderColor="danger700"
						hoverBorderColor="danger700"
					>
						{filSize === undefined ? 'Size' : filSize}
					</Dropdown>
					<Input
						placeholder="Table #"
						name='tableNo'
						type='number'
						w='5.75rem'
						m={{ l: '0.5rem', r: '0.5rem' }}
						borderColor="danger500"
						focusBorderColor="danger700"
						hoverBorderColor="danger700"
						onChange={(e) => this.props.updateFilter('filTableNo', parseInt(e.target.value))}
					/>

					<Button
						rounded="lg"
						left='5rem'
						bg="danger700"
						hoverBg="danger600"
						onClick={() => this.clearAllFilters()}
					>
						Clear All
					</Button>
				</Div>
			</Container >
		);
	}
}