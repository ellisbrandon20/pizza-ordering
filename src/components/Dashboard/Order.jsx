import { Component } from "react";
import { Row, Col, Text, Image, Container, Icon, Button } from "atomize";
import pizza from "../../images/pizza.png";
import { deleteOrder } from "../../apis";

export class Order extends Component {

	deletePizzaOrder = (id) => {
		deleteOrder(id)
			.then(responseData => {
				this.props.changeUpdateOrders(true);
			}).catch(_ => {
				this.props.goToScreen('Error');
			});
	}

	render() {
		const { order } = this.props;

		return (
			<Container
				w="50rem"
				h="8rem"
				m={{ l: 'auto', r: 'auto', t: '1.25rem' }}
				rounded="lg"
				className="transparent-container"
				transform='translateY(5rem)'
			>
				<Row top='50%'>
					<Col size='5'>
						<Row>
							<Col size="12">
								<Text>{`Order ${order.Order_ID}`}</Text>
							</Col>
						</Row>
						<Row>
							<Col size="12">
								<Image src={pizza}
									maxH="5rem"
									maxW="5rem"
								/>
							</Col>
						</Row>
					</Col>

					<Col size="5">
						<Row>
							<Col size="12">
								<Row>
									<Col size="12">
										<Text>
											{`Table ${order.Table_No}`}
										</Text>
									</Col>
								</Row>
								<Row>
									<Col size="12">
										<Text>
											{`Crust: ${order.Crust}`}

										</Text>
									</Col>
								</Row>
								<Row>
									<Col size="12">
										<Text>
											{`Flavor: ${order.Flavor}`}
										</Text>
									</Col>
								</Row>
								<Row>
									<Col size="12">
										<Text>
											{`Size: ${order.Size}`}
										</Text>
									</Col>
								</Row>
							</Col>
						</Row>
					</Col>
					<Col size="2">
						<Button
							h="2.5rem"
							w="2.5rem"
							bg="danger700"
							hoverBg="danger600"
							rounded="circle"
							m={{ r: "1rem" }}
							transform='translateY(100%)'
							onClick={() => this.deletePizzaOrder(order.Order_ID)}
						>
							<Icon name="DeleteSolid" size="20px" color="white" />
						</Button>
					</Col>
				</Row>
			</Container>
		);
	}
}