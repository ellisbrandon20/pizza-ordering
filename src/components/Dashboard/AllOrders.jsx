import { Component } from "react";
import { Order } from "./Order";

// TODO - Handle 0 Orders

export class AllOrders extends Component {
	render() {
		const { orders } = this.props;

		return (
			<>
				{orders.map(order => {
					return (
						<Order key={order.Order_ID} order={order} changeUpdateOrders={this.props.changeUpdateOrders} />
					);
				})}
			</>
		);
	}
}