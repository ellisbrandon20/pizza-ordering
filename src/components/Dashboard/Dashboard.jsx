import { Button, Div, Icon, Notification, Text } from "atomize";
import { Component } from "react";
import { getPizza } from "../../apis";
import { AllOrders } from "./AllOrders";
import { Filter } from "./Filter";

export class Dashboard extends Component {
	constructor(props) {
		super(props);

		this.state = {
			'orders': [],
			'filCrust': undefined,
			'filFlavor': undefined,
			'filSize': undefined,
			'filTableNo': undefined,
			'updateOrders': false,
			'successNotif': false
		}
	}

	componentDidMount() {
		this.getPizzaOrders();
	}

	componentDidUpdate(_, prevState) {
		// call API to get up to date orders when an order is deleted
		if (this.state.updateOrders === true) {
			this.getPizzaOrders();
			this.changeUpdateOrders(false);
		}

		// update list of orders based on filter
		let filteredOrders = this.state.orders;
		let filters = {
			'Crust': this.state.filCrust,
			'Flavor': this.state.filFlavor,
			'Size': this.state.filSize,
			'Table_No': this.state.filTableNo ? parseInt(this.state.filTableNo) : undefined
		};

		if (this.didFilterChange(prevState, this.state)) {
			filteredOrders = filteredOrders.map(order => {
				if ((filters.Crust !== undefined && order.Crust !== filters.Crust)
					|| (filters.Flavor !== undefined && order.Flavor !== filters.Flavor)
					|| (filters.Size !== undefined && order.Size !== filters.Size)
					|| (filters.Table_No !== undefined && order.Table_No !== filters.Table_No)
				) {
					return { ...order, 'filtered': false };
				}
				return { ...order, 'filtered': true };
			});

			// must have if statement otherwise this called forever onload/onupdate
			this.setState({ 'orders': filteredOrders });
		}
	}

	didFilterChange = (prevState, state) => {
		// TODO - find better solution
		return (prevState.filCrust !== state.filCrust)
			|| (prevState.filFlavor !== state.filFlavor)
			|| (prevState.filSize !== state.filSize)
			|| (
				(state.filTableNo !== undefined)
				&& (!(isNaN(prevState.filTableNo) && isNaN(state.filTableNo)))
				&& (parseInt(prevState.filTableNo) !== parseInt(state.filTableNo))
			);

	};

	getPizzaOrders = () => {
		// 'filtered' attribute adds ability to show all orders or just the filtered orders without losing the data
		getPizza()
			.then(responseData => {
				const filteredOrders = responseData.map(order => {
					return { ...order, 'filtered': true };
				});

				this.setState({ 'orders': filteredOrders });
			}).catch(_ => {
				this.props.goToScreen('Error');
			});

	}

	changeFilter = (filterCategory, filterValue) => {
		this.setState({
			[filterCategory]: filterValue
		});
	}

	changeUpdateOrders = (msg) => {
		this.setState({
			'updateOrders': msg,
			'successNotif': true
		});
	}

	updateFilter = (filter, value) => {
		this.setState({
			[filter]: value
		});
	}

	render() {
		const filteredOrders = this.state.orders.filter(order => order.filtered === true)

		return (
			<>
				<Notification
					bg="success700"
					isOpen={this.state.successNotif}
					onClose={() => this.setState({ successNotif: false })}
					prefix={
						<Icon
							name="Success"
							color="white"
							m={{ r: "0.5rem" }}
						/>
					}
				>
					Successfully Deleted Pizza
				</Notification>


				<Div m={{ b: '5rem' }} transform='translateY(5rem)'>
					<Div
						w="50rem"
						h="auto"
						m={{ l: 'auto', r: 'auto' }}
						p={{ t: '0.5rem', b: '2rem' }}
						rounded="lg"
						className="transparent-container dashboard-main"
					>
						<Button
							rounded="lg"
							m={{ l: 'auto', r: 'auto', t: '0.5rem', b: '3rem' }}
							bg="danger700"
							hoverBg="danger600"
							onClick={() => this.props.goToScreen('NewOrder')}
						>
							Start New Order
						</Button>

						{/* Current Orders View */}
						<Text tag="h2" textSize="display1" m={{ b: '2rem' }}>
							Current Orders
						</Text>

						<Filter
							filCrust={this.state.filCrust}
							filFlavor={this.state.filFlavor}
							filSize={this.state.filSize}
							filTableNo={this.state.filTableNo}
							updateFilter={this.updateFilter}
						/>
					</Div>

					<AllOrders
						orders={filteredOrders}
						changeUpdateOrders={this.changeUpdateOrders}
					/>
				</Div>
			</>
		);
	}
}