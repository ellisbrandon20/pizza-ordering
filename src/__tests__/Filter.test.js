import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Filter } from '../components/Dashboard/Filter';
import { Button, Dropdown, Input } from 'atomize';
Enzyme.configure({ adapter: new Adapter() });

const defaultProps = {
	filCrust: 'crust',
	filFlavor: 'flavor',
	filSize: 'size',
	filTableNo: 1,
	updateFilter: jest.fn()
};

test('AllOrders Snapshot', () => {
	const wrapper = renderer.create(
		<Filter {...defaultProps} />
	).toJSON();
	expect(wrapper).toMatchSnapshot();
});

test('render all Dropdown filters', () => {
	const component = shallow(<Filter {...defaultProps} />);
	const dropdowns = component.find(Dropdown);
	expect(dropdowns).toHaveLength(3);
});

test('render Table No filter', () => {
	const component = shallow(<Filter {...defaultProps} />);
	const input = component.find(Input);
	expect(input).toHaveLength(1);
});

test('render Clear All filters Button', () => {
	const component = shallow(<Filter {...defaultProps} />);
	const clearBtn = component.find(Button);
	expect(clearBtn).toHaveLength(1);
});
