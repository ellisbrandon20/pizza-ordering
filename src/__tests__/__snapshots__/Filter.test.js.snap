// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllOrders Snapshot 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <div
      className=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        crust
        <svg
          className=""
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M12,13.5857864 L16.2928932,9.29289322 C16.6834175,8.90236893 17.3165825,8.90236893 17.7071068,9.29289322 C18.0976311,9.68341751 18.0976311,10.3165825 17.7071068,10.7071068 L12.7071068,15.7071068 C12.3165825,16.0976311 11.6834175,16.0976311 11.2928932,15.7071068 L6.29289322,10.7071068 C5.90236893,10.3165825 5.90236893,9.68341751 6.29289322,9.29289322 C6.68341751,8.90236893 7.31658249,8.90236893 7.70710678,9.29289322 L12,13.5857864 Z"
            fill="#505256"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        flavor
        <svg
          className=""
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M12,13.5857864 L16.2928932,9.29289322 C16.6834175,8.90236893 17.3165825,8.90236893 17.7071068,9.29289322 C18.0976311,9.68341751 18.0976311,10.3165825 17.7071068,10.7071068 L12.7071068,15.7071068 C12.3165825,16.0976311 11.6834175,16.0976311 11.2928932,15.7071068 L6.29289322,10.7071068 C5.90236893,10.3165825 5.90236893,9.68341751 6.29289322,9.29289322 C6.68341751,8.90236893 7.31658249,8.90236893 7.70710678,9.29289322 L12,13.5857864 Z"
            fill="#505256"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        size
        <svg
          className=""
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M12,13.5857864 L16.2928932,9.29289322 C16.6834175,8.90236893 17.3165825,8.90236893 17.7071068,9.29289322 C18.0976311,9.68341751 18.0976311,10.3165825 17.7071068,10.7071068 L12.7071068,15.7071068 C12.3165825,16.0976311 11.6834175,16.0976311 11.2928932,15.7071068 L6.29289322,10.7071068 C5.90236893,10.3165825 5.90236893,9.68341751 6.29289322,9.29289322 C6.68341751,8.90236893 7.31658249,8.90236893 7.70710678,9.29289322 L12,13.5857864 Z"
            fill="#505256"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className=""
    >
      <input
        className=""
        d="flex"
        m={
          Object {
            "l": "0.5rem",
            "r": "0.5rem",
          }
        }
        name="tableNo"
        onChange={[Function]}
        placeholder="Table #"
        type="number"
      />
    </div>
    <button
      className=""
      onClick={[Function]}
    >
      Clear All
    </button>
  </div>
</div>
`;
